### PR TITLE
Implement Enhanced Similarity Search Feature Using Cosine Similarity in FastAPI Application

### PR DESCRIPTION
**Pull request by kingsotn**
Hello! I've been working on implementing the semantic search functionality for our application, and I'm happy to say that the basic structure is now in place. However, I believe we need more testing done to ensure that everything is working correctly.

Here's what I've accomplished so far:
- Implemented a new endpoint for semantic search at /search
- Created a new SearchEndpoint router to handle the endpoint
- Defined a new semantic_search function that performs the actual searching
- Modified the encode function to generate embeddings for texts using OpenAI's API
- Added some example data and tests for the semantic_search function

However, there are still several things that need to be tested and implemented:
- Proper error handling for cases where the OpenAI API returns errors or incomplete results
- More comprehensive tests for the semantic_search function, including edge cases and corner cases
- Integration tests to ensure that the entire workflow from end to end works correctly
- Documentation for the new endpoint and function

I would appreciate it if someone could review my changes and provide feedback before we move forward with merging this PR. Thanks!

### CODE REVIEW

File: app/shared_resources.py
Comment (Olivernyu) [inline app/shared_resources.py:1]: Why introduce this new `document_data_store`? All the html data and text are stored inside `page_data_store`, which can be found using the page id of the corresponding document. Can you please change your implementation to use the current set up of the `page_data_store`? You can find reference of the usage and key convention in `savepage` and `summarize`

File: requirements.txt [hunk 1]
```diff
@@ -1,4 +1,5 @@
 annotated-types==0.6.0
+numpy
```
Comment (Olivernyu) [inline requirements.txt]: Please specify version

File: app/routers/search_endpoint.py [hunk 2]
```diff
@@ -0,0 +1,72 @@
...
+
+router = APIRouter()
+_THRESHOLD = 0.80
+
+
+@router.post("/search/{query}")
+async def semantic_search(
+    query: str,
+    ai_provider: str = Query("openai", description="AI provider for semantic search"),
+):
```
Comment (Olivernyu) [inline app/routers/search_endpoint.py]: As per our current implementation of shared resources, the user has to pass the pageid in the endpoint so the backend can find the corresponding text from `page_data_store`

File: app/shared_resources.py
Comment (kingsotn) [reply app/shared_resources.py:1]: okay done. thanks for the help!

File: requirements.txt (hunk 1 above)
Comment (kingsotn) [reply requirements.txt]: removed

**Comment by kingsotn** [comment]
## Add Semantic Search with Cosine Similarity

### Overview
This PR introduces semantic search functionality in our FastAPI app, using cosine similarity to match queries with documents in `page_data_store`.

### Changes
- **New Endpoint**: Added `/semantic_search` for semantic querying.
- **Cosine Similarity**: Implemented `embed_then_compute` to process queries and documents, calculating similarity scores.
- **Testing**: Included tests to validate the new semantic search feature.

File: app/routers/search_endpoint.py [hunk 3]
```diff
@@ -0,0 +1,67 @@
...
+from numpy import dot
+from numpy.linalg import norm
+from ..shared_resources import page_data_store
+
+router = APIRouter()
+_THRESHOLD = 0.80
+
+
+@router.post("/")
+async def semantic_search(
```
Comment (Olivernyu) [inline app/routers/search_endpoint.py]: Can you rename this endpoint to something adjacent to `search_pages` or `find_similar_pages`? This implementation is not performing semantic search, it is a general query to look for similar pages.

**Review by Olivernyu — CHANGES_REQUESTED** [review]

File: app/routers/search_endpoint.py (hunk 3 above)
Comment (kingsotn) [reply app/routers/search_endpoint.py]: Thank you for the feedback. I understand the distinction and agree that the term "semantic search" might not accurately represent the functionality of the endpoint as it currently performs a more general query for similar pages. 

I propose renaming the endpoint to `find_similar_pages`. This name more accurately reflects its purpose of finding pages similar to the input query based on cosine similarity metrics.

I will update the endpoint name, associated tests, and documentation to reflect this change. Your suggestion helps in making the functionality of our API clearer and more intuitive for users.

**Review by Alex-YuYan — APPROVED** [review]
LGTM!

**Review by Olivernyu — APPROVED** [review]
